### DISCUSSION TITLE
"Minimal" wifi firmware for AP mode

### DISCUSSION THREAD

**Discussion by waveform80** [geerlingguy/raspberry-pi-pcie-devices#382] [Ideas]
I've been thoroughly enjoying your videos on various Pi boards (particularly the CM4 carrier boards) and recently noticed something that might be worth bringing to your attention regarding usage of the Pi as a wifi router (my apologies if you've noticed this already; I did a cursory search of the issues but came up empty).

A ["minimal" wifi firmware](https://github.com/RPi-Distro/firmware-nonfree/tree/bullseye/debian/config/brcm80211/cypress) for the 43455 was added to the Pi's [firmware-nonfree](https://github.com/RPi-Distro/firmware-nonfree) repo (which I tend to keep a vague eye on as it's rather important to my day job!). The intended purpose of the minimal firmware is to strip out "extra" functionality in order to increase the number of simultaneous clients that can be supported by the Pi's built-in wifi interface whilst running as an AP.

I doubt it would make an enormous difference to throughput, but perhaps it might be interesting to test nonetheless? One small note: as this is for the 43455 only, it would only be worth testing on boards that use that chip (the 3B+ or the 4B, not the 3B or Zero 2 which use the 43430).

**Comment by geerlingguy** [comment]
Interesting; I'll move this to discussion as it's something I think others could chime in on too. I intend on doing some testing with the CM4's built-in WiFi as an AP at some point soon. I don't have the highest expectations, but having more clients would always be beneficial.

**Reply by waveform80** [reply]
Some additional info from the [Pi Forums](https://forums.raspberrypi.com/viewtopic.php?p=1967056&hilit=wifi+minimal#p1967056):

> I have a few more PixelBlazes though and after 7 connect nothing else can connect to the AP.

In other words, it appears that under some circumstances at least (presumably fairly recent firmware with all the bells & whistles), a Pi AP may be limited to as few as 7 clients.

> It is the RAM inside the WiFi chip that is limited. It only has 800kB and that has to fit the firmware as well.

The minimal firmware theoretically supports as many as 19 clients, albeit with the loss of certain features (roaming, probably not *that* useful in the AP scenario, and dfs-radar which *might* be depending on jurisdiction and other circumstances).

I'm partly interested in gauging whether it would be genuinely useful to have this split out as a separate package to ease the installation (as the README on the firmware notes, subsequent updates can potentially overwrite it currently; that said, it would involve more `dpkg-diverts` ... <*shudder*>). I suspect use of a Pi as an AP is a relatively niche thing anyway, but perhaps that's partly down to existing limitations.
